src/cpu/aarch64/acl_softmax.cpp

```diff
@@ -21,27 +21,147 @@ namespace impl {
 namespace cpu {
 namespace aarch64 {
 
-status_t acl_softmax_fwd_t::execute_forward(const exec_ctx_t &ctx) const {
+const acl_softmax_fwd_t::pd_t *acl_softmax_fwd_t::pd() const {
+    return static_cast<const pd_t *>(primitive_t::pd().get());
+}
+
+status_t acl_softmax_fwd_t::pd_t::init(engine_t *engine) {
+
+    bool ok = is_fwd()
+            && set_default_formats() == status::success
+            // ACL only supports matching src/dst (this must come after
+            // set_default_formats() to handle format_kind::any)
+            && *src_md() == *dst_md()
+            && utils::one_of(
+                    src_md()->data_type, data_type::f32, data_type::f16)
+            && attr()->has_default_values();
+    if (!ok) return status::unimplemented;
+
+    // Get memory desc to find sizes and dims
+    const memory_desc_wrapper src_d(src_md());
+    const data_type_t data_type = src_d.data_type();
+
+    // ACL only supports plain tensors, can be permuted but not blocked
+    if (!src_d.is_plain()) return status::unimplemented;
+
+    // Guards against a 0-sized dimension
+    if (src_d.has_zero_dim()) return status::unimplemented;
+
+    // No scaling
+    asp_.beta = 1;
+
+    asp_.is_logsoftmax = is_logsoftmax();
+
+    // The strides give us the in memory inner size
+    dim_t inner_size_ = src_d.blocking_desc().strides[axis()];
+
+    dim_t axis_size_ = axis_size();
+
+    // The outer size is any left-over dimensions not inner or on the axis
+    dim_t outer_size_ = src_d.nelems() / (inner_size_ * axis_size_);
+
+    // In this context, NHWC tells ACL that the logical and physical
+    // dimensions are the same
+    arm_compute::DataLayout acl_layout = arm_compute::DataLayout::NHWC;
+
+    const arm_compute::DataType acl_data_t
+            = acl_utils::get_acl_data_t(data_type);
+
+    const int threads = dnnl_get_max_threads();
+    if (inner_size_ == 1) {
+        // A rough empirical heuristic created by fitting a polynomial
+        // of the tensor sizes and thread count to the run time of the
+        // ref and ACL softmax. This variable is greater than zero when
+        // ref is faster, and less than zero when ACL is faster. We can
+        // interpret the constant term as the constant overhead
+        // associated with calling the external library and the negative
+        // coefficient on total_size as ACL being faster at processing
+        // each element
+        double acl_ref_performance_diff = 1 + 0.005 * outer_size_
+                - 0.0027 * axis_size_
+                        * std::ceil(double(outer_size_) / threads);
+        if (threads > 1 || outer_size_ > 1) {
+            // Using threads within ACL adds another constant overhead
+            acl_ref_performance_diff += 17;
+        }
```

Comment:
Given this is used in 2 separate spots, maybe put that inside a lambda before the `inner_size_ == 1` conditional?